src/ParseCLP.js

```diff
@@ -0,0 +1,898 @@
+VALID_PERMISSIONS_EXTENDED.set('protectedFields', []);
+
+/**
+ * Creates a new CLP.
+ * If no argument is given, the CLP has no permissions for anyone.
+ * If the argument is a Parse.User, the CLP will have read and write
+ *   permission for only that user.
+ * If the argument is any other JSON object, that object will be interpretted
+ *   as a serialized CLP created with toJSON().
+ *
+ * <p>A CLP, or Class Level Permissions can be added to any
+ * <code>Parse.Schema</code> to restrict access to only a subset of users
+ * of your application.</p>
+ * @alias Parse.CLP
+ */
+class ParseCLP {
+  permissionsMap: PermissionsMap;
+
+  /**
+   * @param {(Parse.User|Parse.Role|Object)} userId The user to initialize the CLP for
+   */
+  constructor(userId: ParseUser | ParseRole | PermissionsMap) {
+
+    this.permissionsMap = {};
+    // Initialize permissions Map with default permissions
+    for (const permission of VALID_PERMISSIONS.entries()) {
+      this.permissionsMap[permission[0]] = Object.assign({}, permission[1]);
+    }
+    // Initialize permissions Map with default extended permissions
+    for (const permission of VALID_PERMISSIONS_EXTENDED.entries()) {
+      this.permissionsMap[permission[0]] = Object.assign(permission[1] instanceof Array ? [] : {}, permission[1]);
+    }
+
+
+    if (userId && typeof userId === 'object') {
+      if (userId instanceof ParseUser) {
+        this.setReadAccess(userId, true);
+        this.setWriteAccess(userId, true);
+        this.setAddAccess(userId, true);
+      } else if (userId instanceof ParseRole) {
+        this.setRoleReadAccess(userId, true);
+        this.setRoleWriteAccess(userId, true);
+        this.setRoleAddAccess(userId, true);
+      } else {
+        for (const permission in userId) {
+          if (userId.hasOwnProperty(permission)) {
+            const users = userId[permission];
+            const isValidPermission = !!VALID_PERMISSIONS.get(permission);
+            const isValidPermissionExtended = !!VALID_PERMISSIONS_EXTENDED.get(permission);
+            if (typeof permission !== 'string' || !(isValidPermission || isValidPermissionExtended)) {
+              throw new TypeError('Tried to create an CLP with an invalid permission type.');
+            }
+
+            for (const user in users) {
+              if (users.hasOwnProperty(user)) {
+                const allowed = users[user];
+                if (typeof user !== 'string') {
+                  throw new TypeError('Tried to create an CLP with an invalid user.');
+                }
+
+                if (typeof allowed !== 'boolean' && !isValidPermissionExtended) {
+                  throw new TypeError('Tried to create an CLP with an invalid permission value.');
+                }
+                this.permissionsMap[permission][user] = allowed;
+              }
+            }
+          }
+        }
+      }
+    } else if (typeof userId === 'function') {
```

Comment:
This can be moved to the top.

Reply:
I believe that will be call 99% with correct parameters, so I do not want waste time doing this validation every call.

Reply:
I believe that will be call 99% with correct parameters, so I do not want waste time doing this validation every call.